drivers/dma/dma_stm32_bdma.c

```diff
@@ -0,0 +1,940 @@
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int bdma_stm32_get_memory_increment(enum dma_addr_adj increment,
+					  uint32_t *ll_increment)
+{
+	switch (increment) {
+	case DMA_ADDR_ADJ_INCREMENT:
+		*ll_increment = LL_BDMA_MEMORY_INCREMENT;
+		break;
+	case DMA_ADDR_ADJ_NO_CHANGE:
+		*ll_increment = LL_BDMA_MEMORY_NOINCREMENT;
+		break;
+	case DMA_ADDR_ADJ_DECREMENT:
+		return -ENOTSUP;
+	default:
+		LOG_ERR("Memory increment error. %d", increment);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int bdma_stm32_get_periph_increment(enum dma_addr_adj increment,
+					  uint32_t *ll_increment)
+{
+	switch (increment) {
+	case DMA_ADDR_ADJ_INCREMENT:
+		*ll_increment = LL_BDMA_PERIPH_INCREMENT;
+		break;
+	case DMA_ADDR_ADJ_NO_CHANGE:
+		*ll_increment = LL_BDMA_PERIPH_NOINCREMENT;
+		break;
+	case DMA_ADDR_ADJ_DECREMENT:
+		return -ENOTSUP;
+	default:
+		LOG_ERR("Periph increment error. %d", increment);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int bdma_stm32_disable_channel(BDMA_TypeDef *bdma, uint32_t id)
+{
+	int count = 0;
+
+	for (;;) {
+		if (stm32_bdma_disable_channel(bdma, id) == 0) {
+			return 0;
+		}
+		/* After trying for 5 seconds, give up */
+		if (count++ > (5 * 1000)) {
+			return -EBUSY;
+		}
+		k_sleep(K_MSEC(1));
+	}
+
+	return 0;
+}
+
+static bool bdma_stm32_is_valid_memory_address(const uint32_t address, const uint32_t size)
+{
+	/* The BDMA can only access memory addresses in SRAM4 */
+
+	const uint32_t sram4_start = DT_REG_ADDR(DT_NODELABEL(sram4));
+	const uint32_t sram4_end = sram4_start + DT_REG_SIZE(DT_NODELABEL(sram4));
+
+	if (address < sram4_start) {
+		return false;
+	}
+
+	if (address + size > sram4_end) {
+		return false;
+	}
+
+	return true;
+}
+
+
+BDMA_STM32_EXPORT_API int bdma_stm32_configure(const struct device *dev,
+					     uint32_t id,
+					     struct dma_config *config)
+{
+	const struct bdma_stm32_config *dev_config = dev->config;
+	struct bdma_stm32_channel *channel =
+				&dev_config->channels[id];
+	BDMA_TypeDef *bdma = (BDMA_TypeDef *)dev_config->base;
+	LL_BDMA_InitTypeDef BDMA_InitStruct;
+	int index;
+	int ret;
+
+	LL_BDMA_StructInit(&BDMA_InitStruct);
+
+	if (id >= dev_config->max_channels) {
+		LOG_ERR("cannot configure the bdma channel %d.", id);
+		return -EINVAL;
+	}
+
+	if (channel->busy) {
+		LOG_ERR("bdma channel %d is busy.", id);
+		return -EBUSY;
+	}
+
+	if (bdma_stm32_disable_channel(bdma, id) != 0) {
+		LOG_ERR("could not disable bdma channel %d.", id);
+		return -EBUSY;
+	}
+
+	bdma_stm32_clear_channel_irq(dev, id);
+
+	/* Check potential BDMA override */
+	if (config->linked_channel == STM32_DMA_HAL_OVERRIDE) {
+		/* BDMA channel is overridden by HAL BDMA
+		 * Retain that the channel is busy and proceed to the minimal
+		 * configuration to properly route the IRQ
+		 */
+		channel->busy = true;
+		channel->hal_override = true;
+		channel->bdma_callback = config->dma_callback;
+		channel->user_data = config->user_data;
+		return 0;
+	}
```

Comment:
Is that a possible case ?
Today `STM32_DMA_HAL_OVERRIDE` happens when QSPI/OSPI/SDMMC are enabling DMA.
But in those cases, they're using DMA channels not BDMA ones

Reply:
True. I removed this `if`.